asadiqbal08/mitxpro 142 - Product Page: What You'll Learn

asadiqbal08 opened the pull request:
### Change Set Includes:

1. Re-structured the CMS course/program pages and used the Parent-Child concept for course/program pages in Wagtail. 
2. Restricted the child page to be add only once in a parent. No more than one child page of same type is allowed.
3. Restricted the user for adding another child page into a child page. 

                             Parent  (Allowed)
                             /  \ 
                            /    \ 
                       Child1  Child2

                             Parent  (Not Allowed) //Child of same type 
                             /  \ 
                            /    \ 
                       Child1  Child1


                        Parent  (Not Allowed)
                          / 
                         /   
                       Child1
                       /
                      /
                  Child 2


4. Removed the default `title` field from UI of wagtail Page class. 
5. Designed the page.
6. Populated it with models data.
7. Moved the test files under a test directory for better structure e.g. `courses/tests/test_models.py` etc. 
8. Added test cases for the corresponding changes. 

Task: #142 

1- Admin will click on **ADD CHILD PAGE** under course/program

<img width="1027" alt="Screen Shot 2019-04-22 at 3 57 34 PM" src="https://user-images.githubusercontent.com/7334669/56497544-67b3ca80-6517-11e9-8225-bd6fb292275b.png">

2- Screen will display the list of defined child pages there. 

<img width="1025" alt="Screen Shot 2019-04-22 at 3 57 09 PM" src="https://user-images.githubusercontent.com/7334669/56497545-67b3ca80-6517-11e9-9f5f-50e16a617526.png">

3- If  user clicks on "Learning outcomes page", 
The following page will be displayed where he/she ca add the Title, Description and list of benefits. 

<img width="1280" alt="Screen Shot 2019-04-23 at 2 12 09 PM" src="https://user-images.githubusercontent.com/7334669/56569147-e0775d00-65d1-11e9-9c76-adcc3a0480cb.png">

4- Publish the page and it becomes the child of Course/Program Page there. 

<img width="1278" alt="Screen Shot 2019-04-23 at 3 00 47 PM" src="https://user-images.githubusercontent.com/7334669/56572518-a78eb680-65d8-11e9-88b6-766265542a60.png">



#### 5- If the child page is added, user cannot add it again. It will not be displayed in child page list. As an example, After adding "What you'll learn" , It will not be displayed in list again. 

6- Course page rendering. 

<img width="1276" alt="Screen Shot 2019-04-24 at 5 23 54 PM" src="https://user-images.githubusercontent.com/7334669/56659167-ddee3380-66b5-11e9-9660-de968f57884c.png">

7- Mobile Screen Shot

<img width="211" alt="Screen Shot 2019-04-24 at 5 27 09 PM" src="https://user-images.githubusercontent.com/7334669/56659325-3fae9d80-66b6-11e9-9e81-9180e5bb808d.png">



#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [x] Mobile width screenshots
  - [x] Tag @ferdi or @pdpinch for review
- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

ahmed-arbisoft commented on cms/models.py:
```diff
@@ -2,16 +2,68 @@
+    def can_create_at(cls, parent):
+        # You can only create one of these page under course / program.
+        return super(LearningOutComesPage, cls).can_create_at(parent) \
+               and parent.get_children().type(LearningOutComesPage).count() == 0
```
Wouldn't it be the same for all types of sub-pages on a course/program? I mean, we'll have one page per section, and given the design a section does not repeat, which means all types like `LearningOutcomesPage` and `WhatYouWillLearnPage` etc will always have a maximum count of 1 per `CoursePage` or `ProgramPage`. Does it make sense to move this into `CourseProgramChildPage`?

asadiqbal08 replied:
we need to check that as I am using the query in which pointing to `LearningOutComesPage` only. 
`parent.get_children().type(LearningOutComesPage).count`

asadiqbal08 commented on cms/models.py:
```diff
@@ -90,4 +144,9 @@ class CoursePage(ProductPage):
     content_panels = [FieldPanel("course")] + ProductPage.content_panels
+
+
+LearningOutComesPage._meta.get_field('slug').default = 'what-you-will-learn'
```
#### NOTE : we will remove it and do it differently.

ahmed-arbisoft commented on cms/models.py:
```diff
@@ -5,13 +5,59 @@
 from wagtail.admin.edit_handlers import FieldPanel, StreamFieldPanel
 from wagtail.core import blocks
-from wagtail.core.models import Page
+from wagtail.core.models import Page, Orderable
```
`Orderable` is not being used, could you remove it?

ahmed-arbisoft commented on cms/models.py:
```diff
@@ -5,13 +5,59 @@
+               and parent.get_children().type(cls).count() == 0
+
+
+class LearningOutComesPage(CourseProgramChildPage):
```
Could you rename it to `LearningOutcomesPage`? (small c in come), otherwise it will get rendered as `Learning Out Comes Page` in the admin description.

ahmed-arbisoft commented on courses/models.py:
```diff
@@ -125,6 +126,12 @@ def duration(self):
+    def outcomes(self):
+        """Gets the learning outcomes from the associated Page children if it exists"""
+        children = self.page.get_children()
+        return children.type(LearningOutComesPage).first().specific
```
How about `children.type(LearningOutcomesPage).first().specific if self.page and self.page.children.type(LearningOutcomesPage).count() else None`

ahmed-arbisoft commented on courses/templates/course_detail.html:
```diff
@@ -42,5 +42,18 @@ <h2>{{ course.subhead }}</h2>
       </div>
     </div>
   </div>
+  <div class="access-world-block">
```
Shouldn't it be `learning-outcomes-block`? This can add to confusion.

ahmed-arbisoft commented on courses/templates/course_detail.html:
```diff
@@ -42,5 +42,18 @@ <h2>{{ course.subhead }}</h2>
   </div>
+  <div class="access-world-block">
+      <div class="container">
+        <div class="info-holder">
```
Something like `title-container` or `heading-container`

ahmed-arbisoft commented on courses/templates/course_detail.html:
```diff
@@ -42,5 +42,18 @@ <h2>{{ course.subhead }}</h2>
+            <h1>{{ course.outcomes.heading | safe }}</h1>
+            <p>{{ course.outcomes.sub_heading | safe }}</p>
+        </div>
+        <ul class="access-list">
```
`learning-outcomes-list`

ahmed-arbisoft commented on mitxpro/settings.py:
```diff
@@ -30,7 +30,7 @@
 ALLOWED_HOSTS = ["*"]
 
-SECURE_SSL_REDIRECT = get_bool("MITXPRO_SECURE_SSL_REDIRECT", True)
+SECURE_SSL_REDIRECT = get_bool("MITXPRO_SECURE_SSL_REDIRECT", False)
```
Do we really need to disable secure SSL redirects here? Or is this a side-effect of rebasing/merging from master or something?

asadiqbal08 replied:
haha. push some un used stuff. let me remove it. Thanks

ahmed-arbisoft commented on static/scss/detail/access-world.scss:
```diff
@@ -0,0 +1,54 @@
+.access-world-block {
```
Rename to something like `learning-outcomes-block`.

ahmed-arbisoft commented on static/scss/layout.scss:
```diff
@@ -10,6 +10,7 @@
 @import "catalog";
 @import "checkout";
 @import "detail/header";
+@import "detail/access-world";
```
Rename to `detail/learning-outcomes`

ahmed-arbisoft commented on static/scss/detail/access-world.scss:
```diff
@@ -0,0 +1,54 @@
+      }
+    }
+  }
+}
```
Missing newline at the end.

ahmed-arbisoft commented on static/scss/detail/access-world.scss:
```diff
@@ -0,0 +1,54 @@
+  background: white url('#{$static-path}/images/bg-access.png') repeat-x 0 0;
+  padding: 50px 0;
+
+  .info-holder {
```
Something like `title-holder` or `heading-container`.

ahmed-arbisoft commented on static/scss/detail/access-world.scss:
```diff
@@ -0,0 +1,54 @@
+    margin: 0 0 30px;
+  }
+
+  .access-list {
```
Rename to `learning-outcomes-list`

asadiqbal08 commented:
I moved the tests files under 'tests' directory and add the *test* word in start of files for better file management and readiability. Moving files in separate test directory would be helpful to identify tests easily.

gsidebo commented:
@asadiqbal08 If that's a change that we want to make in this codebase, we would want to make it for all of the test files. I agree that there are benefits to the separate tests module, but if we're going to change this pattern, I would like there to be at least a short discussion and agreement within the team. In the interest of keeping the codebase consistent, I'd prefer these test files match the pattern of the rest of the codebase

gsidebo commented on static/scss/detail/learning-outcomes.scss:
```diff
@@ -0,0 +1,56 @@
+    font-weight: 600;
+  }
+
+  p {
```
With my suggestion above regarding changing the `p` to an `h3`, this shouldn't be necessary. Can you try the following changes?:

- Add a style definition for `h1, h3`: `margin-top: 0; margin-bottom: 30px;`
- If we really do need the heading to be rich text, add a style to remove margins and padding for `h1 p`

asadiqbal08 replied:
I have added a new file `common.scss` file under `detail` folder that would be used for our own `.scss` files and added the following definition. 

```
body {
  line-height: 28px;
  font-size: 20px;
  font-weight: 500;
}

h1 {
  font-size: 40px;
  line-height: 44px;
  font-weight: 600;
}

h3 {
  font-size: 28px;
  line-height: 38px;
  font-weight: 500;
} 
```

gsidebo commented on static/scss/detail/learning-outcomes.scss:
```diff
@@ -0,0 +1,56 @@
+    text-align: center;
+    font-size: 28px;
+    line-height: 38px;
+    font-weight: 500;
```
Please get rid of `font-size`, `line-height`, and `font-weight` from this definition. The heading and subheading are the only 2 children of the `heading-container` element and they do not seem to share common styles for those properties

gsidebo commented on static/scss/detail/learning-outcomes.scss:
```diff
@@ -0,0 +1,56 @@
+  background: white url('#{$static-path}/images/bg-access.png') repeat-x 0 0;
+  padding: 50px 0;
+
+  .heading-container {
```
There is an existing class applied to other elements called `header-container`. Can you change this class name to match that? Having both is confusing and would make it easy to introduce styling bugs

gsidebo commented on static/scss/detail/learning-outcomes.scss:
```diff
@@ -0,0 +1,56 @@
+      padding: 15px 0;
+      font-size: 24px;
+      line-height: 28px;
+      color: #8a8b8c;
```
`$tile-color-gray` defines this same hex value. Can you use that here and in the line below? While you're at it, there are two spaces before the hex value in the line below

asadiqbal08 commented:
@gsidebo I have reverted the tests structure and renamed to previous state but I would like there should be a story for discussion for it if you agree with that. Thanks.

gsidebo commented:
@asadiqbal08 Just wondering: did you decide to use a child page for the learning outcomes section because the editing view for course/program pages was getting busy with all the editable fields?

asadiqbal08 commented:
@gsidebo There was a discussion with @pdpinch over slack where we decided to use child pages e.g. LearningOutcomes, WhoWillEnroll, LearningTechniques etc for course/program parent pages instead of add all stuff in one page.

asadiqbal08 commented on cms/models.py:
```diff
@@ -12,6 +13,63 @@
+    subpage_types = []
+    heading = RichTextField(
+        blank=False, help_text="Heading highlighting the learning outcomes generally."
+    )
```
no, there is not any specific requirements for that.

asadiqbal08 commented on static/scss/detail/learning-outcomes.scss:
```diff
@@ -0,0 +1,56 @@
+// sass-lint:disable mixins-before-declarations
+
+.learning-outcomes-block {
+  background: white url('#{$static-path}/images/bg-access.png') repeat-x 0 0;
```
I am using the name `dotted-bg`so we can use it in other files too.

ahmed-arbisoft commented:
Can we please merge this and file a bug? That should clear up the rest of
PRs that are backed up because of this.

On Tue, 30 Apr 2019, 8:42 pm Gavin Sidebottom, <notifications@github.com>
wrote:

> *@gsidebo* requested changes on this pull request.
>
> Code looks good, but I'm still getting that white space on the right side
> when that negative margin is applied to the ul. When I remove the
> negative margin, the space goes away. Can you try some different styles to
> provide whatever UI the negative margin is giving you? If it's just not
> possible or it's taking a while to figure out, let me know. We can just
> ignore this and create a bug issue if necessary
>
> [image: ss 2019-04-30 at 11 39 40]
> <https://user-images.githubusercontent.com/14932219/56974533-0904d700-6b3d-11e9-8b7c-1a1d5f076d64.png>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mitodl/mitxpro/pull/170#pullrequestreview-232250241>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKZ74ETZACNMXB5YH4FOLADPTBSG7ANCNFSM4HHWB3ZQ>
> .
>

-- 


The content of this email and attachment(s) are confidential and intended 
solely for the recipient(s) specified in the message. If you received this 
email by error, please alert the sender by replying to this email and 
follow with its deletion, so that we can ensure such an error does not 
occur in future. If you are not the intended recipient, you are strictly 
prohibited to use, copy, disseminate or store this email or its 
attachment(s).

gsidebo commented:
@ahmed-arbisoft That's fine. Are you able to run the python formatter and fix the build? I can do that if you're not able to right now

ahmed-arbisoft commented:
@gsidebo I'm out right now, could you please do it for me? Thanks

On Tue, 30 Apr 2019, 10:41 pm Gavin Sidebottom, <notifications@github.com>
wrote:

> @ahmed-arbisoft <https://github.com/ahmed-arbisoft> That's fine. Are you
> able to run the python formatter and fix the build? I can do that if you're
> not able to right now
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mitodl/mitxpro/pull/170#issuecomment-488046459>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKZ74ETXSYBLCBJTXHKLTOLPTCAE5ANCNFSM4HHWB3ZQ>
> .
>

-- 


The content of this email and attachment(s) are confidential and intended 
solely for the recipient(s) specified in the message. If you received this 
email by error, please alert the sender by replying to this email and 
follow with its deletion, so that we can ensure such an error does not 
occur in future. If you are not the intended recipient, you are strictly 
prohibited to use, copy, disseminate or store this email or its 
attachment(s).

asadiqbal08 commented:
I can do it and merge if gavin is ok with changes.

gsidebo commented:
@asadiqbal08 Please go ahead
